Let me explain in some detail. Consider this program with a single type error.
```
fn main() { let a: u32 = true; }
```

Compile it normally and you get this:
```
error[E0308]: mismatched types
 --> a.rs:1:26
  |
1 | fn main() { let a: u32 = true; }
  |                    ---   ^^^^ expected `u32`, found `bool`
  |                    |
  |                    expected due to this
```

Ok, that's fine.

Now compile it with `-Ztreat-err-as-bug=1` and you get this:
```
error: internal compiler error: `TypeError` when attempting coercion but no error emitted
 --> a.rs:1:26
  |
1 | fn main() { let a: u32 = true; }
  |                          ^^^^

thread 'rustc' panicked at compiler/rustc_errors/src/lib.rs:1467:30:
aborting due to `-Z treat-err-as-bug=1`
```

There is a type error, and the compiler panics afterwards. But the error is labelled as an ICE, there is no error code, the error message is different, and it is missing the "expected" markers. Huh?

Now compile it with `-Ztreat-err-as-bug=2` and you get this:
```
error[E0308]: mismatched types
 --> a.rs:1:26
  |
1 | fn main() { let a: u32 = true; }
  |                    ---   ^^^^ expected `u32`, found `bool`
  |                    |
  |                    expected due to this

thread 'rustc' panicked at compiler/rustc_errors/src/lib.rs:1471:25:
aborting after 1 errors and 1 delayed bugs due to `-Z treat-err-as-bug=2`
```

The type error is back to the original: it's a normal error instead of an ICE, the error message matches, the error code is there, and the "expected" markers are there. And again it panics afterwards. This is the behaviour I would expect with `-Ztreat-err-as-bug`. What is going on?

The final line of the last example explains it: "aborting after 1 errors and 1 delayed bugs". The compiler has "delayed bugs": a kind of second-class diagnostic that only gets printed at the end if no other errors are printed. They're widely used but mostly invisible; you never see them unless there's a bug in the compiler. And yet, `-Ztreat-err-as-bug` treats delayed bugs the same as normal errors.

So now we can now understand the `-Ztreat-err-as-bug=1` output. When handling this type error, the compiler first emits a delayed bug, and then emits a normal error. It's showing the delayed bug, which has a different message, less span information, and happens to be treated as an ICE rather than a normal error followed by a panic.

I think this behaviour is very confusing and not what anybody is likely to want in practice. This PR changes `-Ztreat-err-as-bug` so it ignores delayed bugs, unless they get promoted to proper bugs and are printed.

cc @compiler-errors

--- a/compiler/rustc_errors/src/lib.rs
+++ b/compiler/rustc_errors/src/lib.rs
@@ -865,10 +865,6 @@ impl DiagCtxt {
     /// directly).
     #[track_caller]
     pub fn delayed_bug(&self, msg: impl Into<DiagnosticMessage>) -> ErrorGuaranteed {
-        let treat_next_err_as_bug = self.inner.borrow().treat_next_err_as_bug();
-        if treat_next_err_as_bug {
-            self.bug(msg);
-        }
         DiagnosticBuilder::<ErrorGuaranteed>::new(self, DelayedBug(DelayedBugKind::Normal), msg)
             .emit()
     }
@@ -883,10 +879,6 @@ impl DiagCtxt {
         sp: impl Into<MultiSpan>,
         msg: impl Into<DiagnosticMessage>,
     ) -> ErrorGuaranteed {
-        let treat_next_err_as_bug = self.inner.borrow().treat_next_err_as_bug();
-        if treat_next_err_as_bug {
-            self.span_bug(sp, msg);
-        }
         DiagnosticBuilder::<ErrorGuaranteed>::new(self, DelayedBug(DelayedBugKind::Normal), msg)
             .with_span(sp)
             .emit()
@@ -1259,10 +1251,6 @@ impl DiagCtxtInner {
     }
 
     fn emit_diagnostic(&mut self, mut diagnostic: Diagnostic) -> Option<ErrorGuaranteed> {
-        if matches!(diagnostic.level, Error | Fatal) && self.treat_next_err_as_bug() {
-            diagnostic.level = Bug;
-        }
-
         // The `LintExpectationId` can be stable or unstable depending on when it was created.
         // Diagnostics created before the definition of `HirId`s are unstable and can not yet
         // be stored. Instead, they are buffered until the `LintExpectationId` is replaced by
@@ -1296,6 +1284,12 @@ impl DiagCtxtInner {
                 return None;
             }
             _ => {}
+        }
+
+        // This must come after the possible promotion of `DelayedBug` to
+        // `Error` above.
+        if matches!(diagnostic.level, Error | Fatal) && self.treat_next_err_as_bug() {
+            diagnostic.level = Bug;
         }
 
         if diagnostic.has_future_breakage() {
@@ -1387,20 +1381,14 @@ impl DiagCtxtInner {
     }
 
     fn treat_err_as_bug(&self) -> bool {
-        self.flags.treat_err_as_bug.is_some_and(|c| {
-            self.err_count + self.lint_err_count + self.delayed_bug_count() >= c.get()
-        })
+        self.flags.treat_err_as_bug.is_some_and(|c| self.err_count + self.lint_err_count >= c.get())
     }
 
     // Use this one before incrementing `err_count`.
     fn treat_next_err_as_bug(&self) -> bool {
-        self.flags.treat_err_as_bug.is_some_and(|c| {
-            self.err_count + self.lint_err_count + self.delayed_bug_count() + 1 >= c.get()
-        })
-    }
-
-    fn delayed_bug_count(&self) -> usize {
-        self.span_delayed_bugs.len() + self.good_path_delayed_bugs.len()
+        self.flags
+            .treat_err_as_bug
+            .is_some_and(|c| self.err_count + self.lint_err_count + 1 >= c.get())
     }
 
     fn has_errors(&self) -> bool {
@@ -1412,7 +1400,7 @@ impl DiagCtxtInner {
     }
 
     fn flush_delayed(&mut self, kind: DelayedBugKind) {
-        let (bugs, explanation) = match kind {
+        let (bugs, note1) = match kind {
             DelayedBugKind::Normal => (
                 std::mem::take(&mut self.span_delayed_bugs),
                 "no errors encountered even though `span_delayed_bug` issued",
@@ -1422,6 +1410,7 @@ impl DiagCtxtInner {
                 "no warnings or errors encountered even though `good_path_delayed_bugs` issued",
             ),
         };
+        let note2 = "those delayed bugs will now be shown as internal compiler errors";
 
         if bugs.is_empty() {
             return;
@@ -1447,8 +1436,11 @@ impl DiagCtxtInner {
 
             if i == 0 {
                 // Put the overall explanation before the `DelayedBug`s, to
-                // frame them better (e.g. separate warnings from them).
-                self.emit_diagnostic(Diagnostic::new(Bug, explanation));
+                // frame them better (e.g. separate warnings from them). Also,
+                // make it a note so it doesn't count as an error, because that
+                // could trigger `-Ztreat-err-as-bug`, which we don't want.
+                self.emit_diagnostic(Diagnostic::new(Note, note1));
+                self.emit_diagnostic(Diagnostic::new(Note, note2));
             }
 
             let mut bug =
@@ -1474,22 +1466,12 @@ impl DiagCtxtInner {
 
     fn panic_if_treat_err_as_bug(&self) {
         if self.treat_err_as_bug() {
-            match (
-                self.err_count + self.lint_err_count,
-                self.delayed_bug_count(),
-                self.flags.treat_err_as_bug.map(|c| c.get()).unwrap(),
-            ) {
-                (1, 0, 1) => panic!("aborting due to `-Z treat-err-as-bug=1`"),
-                (0, 1, 1) => panic!("aborting due delayed bug with `-Z treat-err-as-bug=1`"),
-                (count, delayed_count, val) => {
-                    if delayed_count > 0 {
-                        panic!(
-                            "aborting after {count} errors and {delayed_count} delayed bugs due to `-Z treat-err-as-bug={val}`",
-                        )
-                    } else {
-                        panic!("aborting after {count} errors due to `-Z treat-err-as-bug={val}`")
-                    }
-                }
+            let n = self.flags.treat_err_as_bug.map(|c| c.get()).unwrap();
+            assert_eq!(n, self.err_count + self.lint_err_count);
+            if n == 1 {
+                panic!("aborting due to `-Z treat-err-as-bug=1`");
+            } else {
+                panic!("aborting after {n} errors due to `-Z treat-err-as-bug={n}`");
             }
         }
     }

--- a/tests/ui/impl-trait/equality-in-canonical-query.clone.stderr
+++ b/tests/ui/impl-trait/equality-in-canonical-query.clone.stderr
@@ -1,4 +1,6 @@
-error: internal compiler error: no errors encountered even though `span_delayed_bug` issued
+note: no errors encountered even though `span_delayed_bug` issued
+
+note: those delayed bugs will now be shown as internal compiler errors
 
 error: internal compiler error: {OpaqueTypeKey { def_id: DefId(rpit::{opaque#0}), args: [] }: OpaqueTypeDecl { hidden_type: OpaqueHiddenType { span: no-location (#0), ty: Alias(Opaque, AliasTy { args: [], def_id: DefId(foo::{opaque#0}) }) } }}
    |
@@ -6,24 +8,18 @@ error: internal compiler error: {OpaqueTypeKey { def_id: DefId(rpit::{opaque#0})
            
 
 error: internal compiler error: error performing ParamEnvAnd { param_env: ParamEnv { caller_bounds: [], reveal: UserFacing }, value: ProvePredicate { predicate: Binder { value: ProjectionPredicate(AliasTy { args: [FnDef(DefId(rpit), []), ()], def_id: DefId(ops::function::FnOnce::Output) }, Term::Ty(Alias(Opaque, AliasTy { args: [], def_id: DefId(foo::{opaque#0}) }))), bound_vars: [] } } }
-  --> $DIR/equality-in-canonical-query.rs:19:5
+  --> $DIR/equality-in-canonical-query.rs:21:5
    |
 LL |     same_output(foo, rpit);
    |     ^^^^^^^^^^^^^^^^^^^^^^
    |
 
-  --> $DIR/equality-in-canonical-query.rs:19:5
+  --> $DIR/equality-in-canonical-query.rs:21:5
    |
 LL |     same_output(foo, rpit);
    |     ^^^^^^^^^^^^^^^^^^^^^^
-
-
-
-
-
-
 
 query stack during panic:
 end of query stack
-error: aborting due to 3 previous errors
+error: aborting due to 2 previous errors
 

--- a/tests/ui/impl-trait/equality-in-canonical-query.rs
+++ b/tests/ui/impl-trait/equality-in-canonical-query.rs
@@ -1,11 +1,13 @@
 // issue: #116877
 // revisions: sized clone
 //[sized] check-pass
-
 //[clone] known-bug: #108498
 //[clone] failure-status: 101
 //[clone] normalize-stderr-test: "DefId\(.*?\]::" -> "DefId("
-//[clone] normalize-stderr-test: "(?m)note: .*$" -> ""
+//[clone] normalize-stderr-test: "(?m)note: we would appreciate a bug report.*\n\n" -> ""
+//[clone] normalize-stderr-test: "(?m)note: rustc.*running on.*\n\n" -> ""
+//[clone] normalize-stderr-test: "(?m)note: compiler flags.*\n\n" -> ""
+//[clone] normalize-stderr-test: "(?m)note: delayed at.*$" -> ""
 //[clone] normalize-stderr-test: "(?m)^ *\d+: .*\n" -> ""
 //[clone] normalize-stderr-test: "(?m)^ *at .*\n" -> ""
 

--- a/tests/ui/mir/lint/storage-live.stderr
+++ b/tests/ui/mir/lint/storage-live.stderr
@@ -4,11 +4,15 @@ error: internal compiler error: broken MIR in Item(DefId(0:8 ~ storage_live[HASH
    |
 LL |             StorageLive(a);
    |             ^^^^^^^^^^^^^^
+   |
+note: delayed at compiler/rustc_mir_transform/src/lint.rs:97:26 - disabled backtrace
+  --> $DIR/storage-live.rs:22:13
+   |
+LL |             StorageLive(a);
+   |             ^^^^^^^^^^^^^^
 
 aborting due to `-Z treat-err-as-bug=1`
 error: the compiler unexpectedly panicked. this is a bug.
 
 query stack during panic:
-#0 [mir_const] preparing `multiple_storage` for borrow checking
-#1 [mir_promoted] promoting constants in MIR for `multiple_storage`
 end of query stack

--- a/tests/ui/mir/lint/storage-return.rs
+++ b/tests/ui/mir/lint/storage-return.rs
@@ -1,4 +1,4 @@
-// compile-flags: -Zlint-mir -Ztreat-err-as-bug
+// compile-flags: -Zlint-mir -Ztreat-err-as-bug -Zeagerly-emit-delayed-bugs
 // failure-status: 101
 // dont-check-compiler-stderr
 // error-pattern: has storage when returning

--- a/tests/ui/treat-err-as-bug/span_delayed_bug.rs
+++ b/tests/ui/treat-err-as-bug/span_delayed_bug.rs
@@ -1,4 +1,4 @@
-// compile-flags: -Ztreat-err-as-bug
+// compile-flags: -Ztreat-err-as-bug -Zeagerly-emit-delayed-bugs
 // failure-status: 101
 // error-pattern: aborting due to `-Z treat-err-as-bug=1`
 // error-pattern: [trigger_span_delayed_bug] triggering a span delayed bug for testing incremental

--- a/tests/ui/type-alias-impl-trait/rpit_tait_equality_in_canonical_query.current.stderr
+++ b/tests/ui/type-alias-impl-trait/rpit_tait_equality_in_canonical_query.current.stderr
@@ -1,4 +1,6 @@
-error: internal compiler error: no errors encountered even though `span_delayed_bug` issued
+note: no errors encountered even though `span_delayed_bug` issued
+
+note: those delayed bugs will now be shown as internal compiler errors
 
 error: internal compiler error: {OpaqueTypeKey { def_id: DefId(get_rpit::{opaque#0}), args: [] }: OpaqueTypeDecl { hidden_type: OpaqueHiddenType { span: no-location (#0), ty: Alias(Opaque, AliasTy { args: [], def_id: DefId(Opaque::{opaque#0}) }) } }}
    |
@@ -6,24 +8,18 @@ error: internal compiler error: {OpaqueTypeKey { def_id: DefId(get_rpit::{opaque
            
 
 error: internal compiler error: error performing ParamEnvAnd { param_env: ParamEnv { caller_bounds: [], reveal: UserFacing }, value: ProvePredicate { predicate: Binder { value: ProjectionPredicate(AliasTy { args: [FnDef(DefId(get_rpit), []), ()], def_id: DefId(ops::function::FnOnce::Output) }, Term::Ty(Alias(Opaque, AliasTy { args: [], def_id: DefId(Opaque::{opaque#0}) }))), bound_vars: [] } } }
-  --> $DIR/rpit_tait_equality_in_canonical_query.rs:28:5
+  --> $DIR/rpit_tait_equality_in_canonical_query.rs:31:5
    |
 LL |     query(get_rpit);
    |     ^^^^^^^^^^^^^^^
    |
 
-  --> $DIR/rpit_tait_equality_in_canonical_query.rs:28:5
+  --> $DIR/rpit_tait_equality_in_canonical_query.rs:31:5
    |
 LL |     query(get_rpit);
    |     ^^^^^^^^^^^^^^^
-
-
-
-
-
-
 
 query stack during panic:
 end of query stack
-error: aborting due to 3 previous errors
+error: aborting due to 2 previous errors
 

--- a/tests/ui/type-alias-impl-trait/rpit_tait_equality_in_canonical_query.rs
+++ b/tests/ui/type-alias-impl-trait/rpit_tait_equality_in_canonical_query.rs
@@ -12,7 +12,10 @@
 //[current] known-bug: #108498
 //[current] failure-status: 101
 //[current] normalize-stderr-test: "DefId\(.*?\]::" -> "DefId("
-//[current] normalize-stderr-test: "(?m)note: .*$" -> ""
+//[current] normalize-stderr-test: "(?m)note: we would appreciate a bug report.*\n\n" -> ""
+//[current] normalize-stderr-test: "(?m)note: rustc.*running on.*\n\n" -> ""
+//[current] normalize-stderr-test: "(?m)note: compiler flags.*\n\n" -> ""
+//[current] normalize-stderr-test: "(?m)note: delayed at.*$" -> ""
 //[current] normalize-stderr-test: "(?m)^ *\d+: .*\n" -> ""
 //[current] normalize-stderr-test: "(?m)^ *at .*\n" -> ""
 

